File: app/src/at-phuongle/java/com/asiantech/intern20summer1/intro/IslandOfKnowledge.java
package com.asiantech.intern20summer1.intro;

import java.util.ArrayList;
import java.util.List;

public class IslandOfKnowledge {
    /**
     * Write a function that checks if you and your friend are equally strong.
     */
    boolean areEquallyStrong(int yourLeft, int yourRight, int friendsLeft, int friendsRight) {
        int myHeaviestWeights = (yourLeft > yourRight) ? yourLeft : yourRight;
        int friendsHeaviestWeights = (friendsLeft > friendsRight) ? friendsLeft : friendsRight;
        int mySum = yourLeft + yourRight;
        int friendsSum = friendsLeft + friendsRight;
        if (mySum == friendsSum) {
            if (myHeaviestWeights == friendsHeaviestWeights) {
                return true;
            }
        }
        return false;
    }

    /**
     * Write a function that returns the maximal absolute difference between any two of its adjacent elements.
     */
    int arrayMaximalAdjacentDifference(int[] inputArray) {
        int maxDiff = 0;
        for (int i = 0; i < inputArray.length - 1; i++) {
            int diff = Math.abs(inputArray[i] - inputArray[i + 1]);
            if (maxDiff < diff) {
                maxDiff = diff;
            }
        }
        return maxDiff;
    }

    /**
     * Write a function that validates IPv4 string
     */
    boolean isIPv4Address(String inputString) {
        String[] numbers = inputString.split("[.]");
        if (numbers.length != 4) {
            return false;
        } else {
            for (int i = 0; i < numbers.length; i++) {
                try {
                    if ((numbers[i].length() == 2) && Integer.parseInt(numbers[i]) < 10) {
                        return false;
                    }
                    if (numbers[i].isEmpty()) {
                        return false;
                    }
                    if (Integer.parseInt(numbers[i]) < 0 || Integer.parseInt(numbers[i]) > 255) {
                        return false;
                    }
                } catch (Exception ex1) {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Write a function that returns the minimal length of the jump enough to avoid all the obstacles.
     */
    int avoidObstacles(int[] inputArray) {
        boolean check = false;
        int count = 1;
        while (true) {
            for (int i = 0; i < inputArray.length; i++) {
                if (inputArray[i] % count != 0) {
                    check = true;
                } else {
                    check = false;
                    break;
                }
            }
            if (check == true) {
                return count;
            }
            count++;
        }
    }

    /**
     * Write a function that blurs images by Box Blur algorithm.
     */
    int[][] boxBlur(int[][] image) {
        int sum = 0;
        int count = 0;
        List<Integer> list = new ArrayList<>();

        for (int i = 0; i < image.length; i++) {
            for (int j = 0; j < image[0].length; j++) {
                count += 4;
                if (i < (image.length - 1)) {
                    count += 2;
                }
                if (j < (image[0].length - 1)) {

                    count += 2;
                }
                if (i < (image.length - 1) && j < (image[0].length - 1)) {

                    count += 1;
                }
                try {
                    if (count == 9) {
                        sum += image[i][j];
                        sum += image[i][j - 1];
                        sum += image[i - 1][j];
                        sum += image[i - 1][j - 1];
                        sum += image[i + 1][j];
                        sum += image[i + 1][j - 1];
                        sum += image[i][j + 1];
                        sum += image[i - 1][j + 1];
                        sum += image[i + 1][j + 1];

                        list.add(sum / count);
                    }
                } catch (Exception ex2) {
                    //Do something
                }
                sum = 0;
                count = 0;
            }
        }

        int[][] blurImage = new int[image.length - 2][image[0].length - 2];
        for (int i = 0; i < blurImage.length; i++) {
            for (int j = 0; j < blurImage[0].length; j++) {
                blurImage[i][j] = list.get(0);
                list.remove(0);
            }
        }

        return blurImage;
    }

    /**
     * Write a function that returns rectangular matrix of the same size as matrix each cell of which
     * contains an integer equal to the number of mines in the neighboring cells.
     */
    int[][] minesweeper(boolean[][] matrix) {
        int[][] result = new int[matrix.length][matrix[0].length];
        for (int i = 0; i < result.length; i++) {
            for (int j = 0; j < result[0].length; j++) {
                result[i][j] = 0;
            }
        }

        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                if (matrix[i][j]) {
                    try {
                        if (j > 0) result[i][j - 1]++;
                        if (j < matrix[0].length - 1) result[i][j + 1]++;
                        if (i > 0) result[i - 1][j]++;
                        if (i < matrix.length - 1) result[i + 1][j]++;
                        if (i > 0 && j > 0) result[i - 1][j - 1]++;
                        if (i < matrix.length - 1 && j > 0) result[i + 1][j - 1]++;
                        if (i > 0 && j < matrix[0].length - 1) result[i - 1][j + 1]++;
                        if (i < matrix.length - 1 && j < matrix[0].length - 1)
                            result[i + 1][j + 1]++;
                    } catch (Exception ex3) {
                        //Do something
                    }
                }
            }
        }

        return result;
    }
}

File: app/src/at-phuongle/java/com/asiantech/intern20summer1/kotlin/IslandOfKnowledge.kt
package com.asiantech.intern20summer1.kotlin

object IslandOfKnowledge {
    private const val FOUR = 4
    private const val NINE = 9
    private const val TEN = 10
    private const val MAX_IPv4 = 255

    fun areEquallyStrong(
        yourLeft: Int,
        yourRight: Int,
        friendsLeft: Int,
        friendsRight: Int
    ): Boolean {
        var myHeaviestWeights: Int
        var friendsHeaviestWeights: Int
        var mySum: Int = yourLeft + yourRight
        var friendsSum: Int = friendsLeft + friendsRight

        if (yourLeft > yourRight) {
            myHeaviestWeights = yourLeft
        } else {
            myHeaviestWeights = yourRight
        }

        if (friendsLeft > friendsRight) {
            friendsHeaviestWeights = friendsLeft
        } else {
            friendsHeaviestWeights = friendsRight
        }

        if (mySum == friendsSum) {
            if (myHeaviestWeights == friendsHeaviestWeights) {
                return true;
            }
        }
        return false;
    }

    fun arrayMaximalAdjacentDifference(inputArray: Array<Int>): Int {
        var maxDiff: Int = 0;

        for (i in 0 until inputArray.size - 1) {
            var diff: Int = Math.abs(inputArray[i] - inputArray[i + 1]);
            if (maxDiff < diff) {
                maxDiff = diff;
            }
        }

        return maxDiff;
    }

    fun isIPv4Address(inputString: String): Boolean {
        var check: Boolean = true
        var regex: String = "[.]"
        var numbers: Array<String> = inputString.split(regex.toRegex()).toTypedArray()

        if (numbers.size != FOUR) {
            check = false
        } else {
            for (i in numbers.indices) {
                try {
                    if ((numbers[i].length == 2) && Integer.parseInt(numbers[i]) < TEN) {
                        check = false
                    }
                    if (numbers[i].isNullOrEmpty()) {
                        check = false
                    }
                    if (Integer.parseInt(numbers[i]) < 0 || Integer.parseInt(numbers[i]) > MAX_IPv4) {
                        check = false
                    }
                } catch (ex1: IndexOutOfBoundsException) {
                    check = false
                }
            }
        }

        return check
    }

    fun avoidObstacles(inputArray: Array<Int>): Int {
        var check: Boolean = false
        var count: Int = 1
        while (true) {
            for (i in 0 until inputArray.size) {
                if (inputArray[i] % count != 0) {
                    check = true
                } else {
                    check = false
                    break
                }
            }

            if (check == true) {
                break
            }

            count++;
        }

        return count
    }

    fun boxBlur(image: Array<Array<Int>>): Array<IntArray> {
        var sum: Int = 0
        var count: Int = 0
        var list: MutableList<Int> = mutableListOf()

        for (i in image.indices) {
            for (j in image[0].indices) {
                count += FOUR

                if (i < (image.size - 1)) {
                    count += 2
                }

                if (j < (image[0].size - 1)) {
                    count += 2
                }

                if (i < (image.size - 1) && j < (image[0].size - 1)) {
                    count += 1
                }

                try {
                    if (count == NINE) {
                        sum += image[i][j]
                        sum += image[i][j - 1]
                        sum += image[i - 1][j]
                        sum += image[i - 1][j - 1]
                        sum += image[i + 1][j]
                        sum += image[i + 1][j - 1]
                        sum += image[i][j + 1]
                        sum += image[i - 1][j + 1]
                        sum += image[i + 1][j + 1]

                        list.add(sum / count)
                    }
                } catch (ex2: Exception) {
                    println("Exception 2")
                }

                sum = 0
                count = 0
            }
        }

        var blurImage: Array<IntArray> = Array(image.size - 2) { IntArray(image[0].size - 2) }
        for (element in blurImage) {
            for (j in blurImage[0].indices) {
                element[j] = list[0]
                list.remove(0)
            }
        }

        return blurImage
    }

    fun minesweeper(matrix: Array<Array<Boolean>>): Array<IntArray> {
        var result: Array<IntArray> = Array(matrix.size) { IntArray(matrix[0].size) }

        for (element in result) {
            for (j in result[0].indices) {
                element[j] = 0
            }
        }

        for (i in matrix.indices) {
            for (j in matrix[0].indices) {
                if (matrix[i][j]) {
                    try {
                        if (j > 0) result[i][j - 1]++
                        if (j < matrix[0].size - 1) result[i][j + 1]++
                        if (i > 0) result[i - 1][j]++
                        if (i < matrix.size - 1) result[i + 1][j]++
                        if (i > 0 && j > 0) result[i - 1][j - 1]++
                        if (i < matrix.size - 1 && j > 0) result[i + 1][j - 1]++
                        if (i > 0 && j < matrix[0].size - 1) result[i - 1][j + 1]++
                        if (i < matrix.size - 1 && j < matrix[0].size - 1) result[i + 1][j + 1]++
                    } catch (ex3: Exception) {
                        println("Exception 3")
                    }
                }
            }
        }

        return result
    }
}
